### ISSUE TITLE
[Feature Request] Support custom, tensor-based decay in TD(lambda)

### ISSUE THREAD

**Issue by vmoens** [facebookresearch/rl#355]
## Motivation

Currently, the vectorised TD(lambda) function only supports scalar gamma values.
However, in some cases we'll want to provide a gamma vector: for instance, the [`MultiStep`](https://github.com/facebookresearch/rl/blob/d295fae9f532862d60e28ca1f509ad1183cf6a56/torchrl/data/postprocs/postprocs.py#L109) class computes a gamma tensor of the same size as the reward that is to be used to compute decayed reward sums.

## Solution

The proposed solution is to allow [`_custom_conv1d`](https://github.com/facebookresearch/rl/blob/266ddff05ce0c31ba526c0124e0f025ff5273286/torchrl/objectives/returns/functional.py#L167) to receive filters (i.e. stack of gamma values) with a certain batch size. The conv1d operation does not support batching but we can do something like this:

```python
if filter.ndimension() > 2:  # then filter has a batch size
    filter = filter.reshape(batch_size, 1, 1, T)
    functorch.vmap(torch.conv1d, (None, 0))(val_pad, filter)  # (None, 0) tells vmap to expand val_pad and not filter
else:
   do_as_before()
```

In this implementation, `functorch` will do the vectorised map of filter over val_pad (i.e. expand val_pad and execute conv1d over it).

Naturally, the creation of the filter and the reshaping will need to be adapted accordingly.

Tests should be added to account for this feature and placed in `test_costs.py`.
For instance:

```python
batch = 10
time = 11
next_value = torch.randn(batch, time, 1)
done = torch.zeros(batch, time, 1, dtype=torch.bool)
reward = torch.randn(batch, time, 1)

gamma = 0.99
lmbda = 0.95

gamma_tensor = torch.full(gamma, (batch, time, 1))

v1 = vec_td_lambda_return_estimate(gamma, lmbda, next_state_value, reward, done)
v2 = vec_td_lambda_return_estimate(gamma_tensor, lmbda, next_state_value, reward, done)

assert_isclose(v1, v2)

# same with last done being true
done[:, -1] = True  # terminating trajectory
gamma_tensor[:, -1] = 0.0

v1 = vec_td_lambda_return_estimate(gamma, lmbda, next_state_value, reward, done)
v2 = vec_td_lambda_return_estimate(gamma_tensor, lmbda, next_state_value, reward, done)

assert_isclose(v1, v2)

```
